File: Sources/ClipDeckKit/Model/Template.swift
import Foundation

/// Um template de mensagem.
public struct Template: Codable, Identifiable, Equatable, Hashable {
    public let id: UUID
    public var title: String
    public var body: String
    public var folderID: UUID?
    public var tags: [String]

    /// Quantas vezes foi usado. Alimenta o ranking: o que você mais usa sobe sozinho.
    public var useCount: Int
    public var lastUsedAt: Date?
    public var createdAt: Date
    public var updatedAt: Date

    public init(
        id: UUID = UUID(),
        title: String,
        body: String,
        folderID: UUID? = nil,
        tags: [String] = [],
        useCount: Int = 0,
        lastUsedAt: Date? = nil,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.title = title
        self.body = body
        self.folderID = folderID
        self.tags = tags
        self.useCount = useCount
        self.lastUsedAt = lastUsedAt
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    /// Primeira linha não vazia do corpo, para preview na lista.
    public var preview: String {
        body
            .split(separator: "\n", omittingEmptySubsequences: true)
            .first
            .map { $0.trimmingCharacters(in: .whitespaces) }
            ?? ""
    }

    /// Decodificação tolerante: campos novos adicionados em versões futuras não
    /// invalidam um arquivo antigo, e um arquivo editado à mão sem `tags` ou
    /// `useCount` continua carregando em vez de derrubar a biblioteca inteira.
    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decodeIfPresent(UUID.self, forKey: .id) ?? UUID()
        title = try c.decodeIfPresent(String.self, forKey: .title) ?? "Sem título"
        body = try c.decodeIfPresent(String.self, forKey: .body) ?? ""
        folderID = try c.decodeIfPresent(UUID.self, forKey: .folderID)
        tags = try c.decodeIfPresent([String].self, forKey: .tags) ?? []
        useCount = try c.decodeIfPresent(Int.self, forKey: .useCount) ?? 0
        lastUsedAt = try c.decodeIfPresent(Date.self, forKey: .lastUsedAt)
        createdAt = try c.decodeIfPresent(Date.self, forKey: .createdAt) ?? Date()
        updatedAt = try c.decodeIfPresent(Date.self, forKey: .updatedAt) ?? Date()
    }
}

File: windows/src/ClipDeck.Core/Models/Template.cs
using System.Text.Json.Serialization;

namespace ClipDeck.Core.Models;

/// <summary>Um template de mensagem.</summary>
public sealed class Template
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string Title { get; set; } = "Sem título";
    public string Body { get; set; } = string.Empty;
    public Guid? FolderId { get; set; }
    public List<string> Tags { get; set; } = new();

    /// <summary>Quantas vezes foi usado. Alimenta o ranking da busca.</summary>
    public int UseCount { get; set; }

    public DateTimeOffset? LastUsedAt { get; set; }
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.Now;
    public DateTimeOffset UpdatedAt { get; set; } = DateTimeOffset.Now;

    /// <summary>Primeira linha não vazia do corpo, para preview na lista.</summary>
    [JsonIgnore]
    public string Preview
    {
        get
        {
            foreach (var line in Body.Split('\n'))
            {
                var trimmed = line.Trim();
                if (trimmed.Length > 0) return trimmed;
            }
            return string.Empty;
        }
    }

    public Template Clone() => new()
    {
        Id = Id,
        Title = Title,
        Body = Body,
        FolderId = FolderId,
        Tags = new List<string>(Tags),
        UseCount = UseCount,
        LastUsedAt = LastUsedAt,
        CreatedAt = CreatedAt,
        UpdatedAt = UpdatedAt,
    };
}
